src/scopeck.rs

```diff
@@ -151,156 +151,148 @@ struct ScopeState<'a> {
     frames_out: Vec<Frame>,
 }
 
-fn push_diagnostic(state: &mut ScopeState, ix: StatementIndex, diag: Diagnostic) {
-    state.diagnostics.entry(ix).or_insert(Vec::new()).push(diag);
-}
+impl<'a> ScopeState<'a> {
+    fn push_diagnostic(&mut self, ix: StatementIndex, diag: Diagnostic) {
```

Comment:
Right, you're thinking in Java.  This is not an object-oriented language and the project got a lot smoother once I stopped pretending that it was.  Early versions of smm3 used `self` a lot more than current versions.  One downside of OO thinking is that it encourages people to put all their fields in one struct; I did, in fact, try to combine ScopeState and ScopeResult once, it cannot work due to lifetimes, but I realized my thinking was the problem.